### PR TITLE
Add new option to `bundle gem` for choosing a CI sevice

### PR DESCRIPTION
**Pull request by hsbt**
# Description:

Migrated from https://github.com/rubygems/bundler/pull/7627

/cc @colby-swandale @indirect 

______________

# Tasks:

- [ ] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: bundler/lib/bundler/cli.rb [hunk 1]
```diff
@@ -570,6 +570,7 @@ def viz
     method_option :rubocop, :type => :boolean, :desc => "Add rubocop to the generated Rakefile and gemspec. Set a default with `bundle config set gem.rubocop true`."
     method_option :test, :type => :string, :lazy_default => "rspec", :aliases => "-t", :banner => "rspec",
                          :desc => "Generate a test directory for your library, either rspec, minitest or test-unit. Set a default with `bundle config set gem.test rspec`."
+    method_option :ci,   :type => :string, :desc => "Generate CI configuration, either Github Actions, Travis CI, Gitlab CI or Circle CI. Set a default with `bundle config set gem.ci (github|travis|gitlab|circle)`"
```
Comment (olleolleolle) [inline bundler/lib/bundler/cli.rb]:
```suggestion
    method_option :ci,   :type => :string, :desc => "Generate CI configuration, either GitHub Actions, Travis CI, GitLab CI or CircleCI. Set a default with `bundle config set gem.ci (github|travis|gitlab|circle)`"
```

This suggestion: I changed to the "fancy" spellings used by the services themselves on their websites.

File: bundler/lib/bundler/cli/gem.rb [hunk 2]
```diff
@@ -231,6 +241,35 @@ def ask_and_set_test_framework
       test_framework
     end
 
+    def ask_and_set_ci
+      ci_template = options[:ci] || Bundler.settings["gem.ci"]
+
+      if ci_template.nil?
+        Bundler.ui.confirm "Do you want to set up automated testing for your gem? " \
+          "Continuous integration services make it easy to see if pull requests have passing tests " \
+          "before you merge them. Bundler supports these services:"\
```
Comment (olleolleolle) [inline bundler/lib/bundler/cli/gem.rb]:
```suggestion
          "before you merge them. Bundler supports these services:" \
```

(Style: Uniform space-after-the-string.)

File: bundler/lib/bundler/cli/gem.rb [hunk 3]
```diff
@@ -231,6 +241,35 @@ def ask_and_set_test_framework
...
+          "* Circle CI:      https://circleci.com/\n" \
```
Comment (olleolleolle) [inline bundler/lib/bundler/cli/gem.rb]:
```suggestion
          "* CircleCI:      https://circleci.com/\n" \
```

File: bundler/lib/bundler/cli/gem.rb [hunk 4]
```diff
@@ -231,6 +241,35 @@ def ask_and_set_test_framework
...
+          "* Github Actions: https://github.com/features/actions\n" \
```
Comment (olleolleolle) [inline bundler/lib/bundler/cli/gem.rb]:
```suggestion
          "* GitHub Actions: https://github.com/features/actions\n" \
```

File: bundler/lib/bundler/cli/gem.rb [hunk 5]
```diff
@@ -231,6 +241,35 @@ def ask_and_set_test_framework
...
+          "* Gitlab CI:      https://docs.gitlab.com/ee/ci/\n" \
```
Comment (olleolleolle) [inline bundler/lib/bundler/cli/gem.rb]:
```suggestion
          "* GitLab CI:      https://docs.gitlab.com/ee/ci/\n" \
```

File: bundler/spec/commands/newgem_spec.rb [hunk 6]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+
+        expect(bundled_app("#{gem_name}/.github/workflows/main.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.travis.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.gitlab-ci.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.circleci/config.yml")).to_not exist
+      end
+    end
+
+    context "--ci set to github" do
+      it "generates a Github Actions config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a GitHub Actions config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 7]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "gem #{gem_name} --ci=github"
+
+        expect(bundled_app("#{gem_name}/.github/workflows/main.yml")).to exist
+      end
+    end
+
+    context "--ci set to gitlab" do
+      it "generates a Gitlab Ci config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a GitLab Ci config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 8]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "gem #{gem_name} --ci=gitlab"
+
+        expect(bundled_app("#{gem_name}/.gitlab-ci.yml")).to exist
+      end
+    end
+
+    context "--ci set to circle" do
+      it "generates a Circle Ci config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a CircleCI config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 9]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "gem #{gem_name} --ci=circle"
+
+        expect(bundled_app("#{gem_name}/.circleci/config.yml")).to exist
+      end
+    end
+
+    context "--ci set to travis" do
+      it "generates a Travis Ci config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a Travis CI config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 10]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "gem #{gem_name} --ci=travis"
+
+        expect(bundled_app("#{gem_name}/.travis.yml")).to exist
+      end
+    end
+
+    context "gem.ci setting set to none" do
+      it "doesnt generate any CI config" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "doesn't generate any CI config" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 11]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        expect(bundled_app("#{gem_name}/.github/workflows/main.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.travis.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.gitlab-ci.yml")).to_not exist
+        expect(bundled_app("#{gem_name}/.circleci/config.yml")).to_not exist
+      end
+    end
+
+    context "gem.ci setting set to github" do
+      it "generates a Github Actions config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a GitHub Actions config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 12]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "config set gem.ci github"
+        bundle "gem #{gem_name}"
+
+        expect(bundled_app("#{gem_name}/.github/workflows/main.yml")).to exist
+      end
+    end
+
+    context "gem.ci setting set to travis" do
+      it "generates a Travis config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a Travis CI config file" do
```

File: bundler/spec/commands/newgem_spec.rb [hunk 13]
```diff
@@ -558,6 +558,94 @@ def create_temporary_dir(dir)
...
+        bundle "config set gem.ci travis"
+        bundle "gem #{gem_name}"
+
+        expect(bundled_app("#{gem_name}/.travis.yml")).to exist
+      end
+    end
+
+    context "gem.ci setting set to gitlab" do
+      it "generates a Gitlab config file" do
```
Comment (olleolleolle) [inline bundler/spec/commands/newgem_spec.rb]:
```suggestion
      it "generates a GitLab CI config file" do
```

**Comment by simi** [comment]
> (Future idea: For services that are not currently represented, would it be possible for them to build a plugin with their template in it?)

It would be useful also for services currently supported to prepare custom template (with some common custom setup, for example setup for private gems). Maybe some entry in `.gemrc` (since bundler is part of rubygems now, right?) specifying:

```yml
ci_templates:
  my_awesome_ci:
    "setup.yml": "~/setup.tt.yml"
```

```bash
bundle gem my_new_awesome_gem --ci=my_awesome_ci
```

I can try to open this once this is merged.

_PS: What about to alias `bundle gem` to `gem init` now as well since the merge?_

**Comment by deivid-rodriguez** [comment]
@simi Note that for now we have only merged the source code repos. Rubygems and bundler are still separate projects. Hopefully we'll be able to merge them at some point though.

**Comment by simi** [comment]
@deivid-rodriguez ahh, there's `.bundle/config` then as well.

**Comment by deivid-rodriguez** [comment]
@FTLam11 Pinging you since you've recently gone through all this code, in case you're interested in getting this over the line. If not, sorry for the ping :pray:.

**Comment by FTLam11** [comment]
@deivid-rodriguez Sure, I'll see what I can do.

**Comment by FTLam11** [comment]
I made changes in my fork of rubygems, and included the suggestions made by @olleolleolle. I don't know if there's an easy way to change the target branch for this PR, should I open another PR and reference this one?

**Comment by deivid-rodriguez** [comment]
You could open PRs targetting the `colby/bundle-gem-ci` branch, but I believe it's easier to open a separate PR superseding this one :+1:

**Comment by hsbt** [comment]
Closing favor of #3667

**Pull request closed without merging**